### PR TITLE
Influxdb indy support

### PR DESCRIPTION
**Pull request by LikeTheSalad**
Part of: https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/11457

These changes replace the use of `@Adive.Local`s and `readOnly = false` to make this instrumentation [indy compatible](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/11546/files#diff-c01c02c7d6cde11986ed4c95f749da6318b0deb89c371d866319371cc1757bd0).

There are a couple of extra changes brought from [this other PR](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/11578) that allow instrumentations already using `@Advice.AssignReturned.*` annotations to work well.

### CODE REVIEW

File: instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java [hunk 1]
```diff
@@ -70,89 +70,71 @@ public void transform(TypeTransformer transformer) {
...
-        @Advice.Argument(0) Query query,
-        @Advice.AllArguments(readOnly = false, typing = Assigner.Typing.DYNAMIC) Object[] arguments,
-        @Advice.FieldValue(value = "retrofit") Retrofit retrofit,
-        @Advice.Local("otelCallDepth") CallDepth callDepth,
-        @Advice.Local("otelRequest") InfluxDbRequest influxDbRequest,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
-      callDepth = CallDepth.forClass(InfluxDBImpl.class);
+    @Advice.AssignReturned.ToAllArguments(index = 0, typing = Assigner.Typing.DYNAMIC)
+    public static Object[] onEnter(
```
Comment (trask) [inline instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java:74]: instead of returning `Object[]`, I like the approach in #11553 of introducing a new class (`HibernateOperationScope` in that PR)

File: instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java [hunk 2]
```diff
@@ -70,89 +70,71 @@ public void transform(TypeTransformer transformer) {
...
+        @Advice.Thrown Throwable throwable, @Advice.Enter Object[] enterArgs) {
+      CallDepth callDepth = CallDepth.forClass(InfluxDBImpl.class);
       if (callDepth.decrementAndGet() > 0) {
         return;
       }
 
-      if (scope == null) {
-        return;
-      }
+      Object extraObject = enterArgs[enterArgs.length - 1];
```
Comment (laurit) [inline instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java:117]: In my opinion this is hard to follow. The size of the array could be either 1 or 2. If it is 2 then the last element is `InfluxDbScope`, if it is 1 then it is an object array. Perhaps it would be better to be more explicit.
```
if (enterArgs != null && enterArgs.length >= 2 && enterArgs[1] instanceof InfluxDbScope) {
  ((InfluxDbScope) enterArgs[1]).close(throwable);
}
````
instead of having a static `InfluxDbScope.end()` you could have an instance method. You don't need to pass `instrumenter()` to `InfluxDbScope`, it can call `instrumenter()` itself.

File: instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java [hunk 3]
```diff
@@ -70,89 +70,71 @@ public void transform(TypeTransformer transformer) {
...
+      InfluxDbScope.end(extraObject, instrumenter(), throwable);
     }
   }
 
   @SuppressWarnings("unused")
   public static class InfluxDbModifyAdvice {
 
     @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static void onEnter(
+    public static Object onEnter(
```
Comment (laurit) [inline instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java:127]: I think now we could return `InfluxDbScope` instead of `Object`

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceTransformer.java [hunk 4]
```diff
@@ -554,105 +583,115 @@ public void visitCode() {
...
-            || returnType.getSort() == Type.OBJECT
-            || returnType.getSort() == Type.ARRAY)) {
-      context.disableReturnTypeChange();
-    }
-    // context is shared by enter and exit advice, if entry advice was rejected don't attempt to
-    // rewrite usages of @Advice.Enter in the exit advice
-    if (!context.canChangeReturnType()) {
-      enterArgument = null;
-    }
+    if (!justDelegateAdvice) {
```
Comment (laurit) [inline javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceTransformer.java:594]: instead of having an if block that surrounds everything could extract the transform that changes the inline advice to delegating advice into helper method and call it here or perhaps even call it instead of `instrument` when only changing the advice type is desired

File: instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java [hunk 5]
```diff
@@ -70,89 +70,71 @@ public void transform(TypeTransformer transformer) {
...
-        @Advice.Local("otelScope") Scope scope) {
-      callDepth = CallDepth.forClass(InfluxDBImpl.class);
+    @Advice.AssignReturned.ToAllArguments(index = 0, typing = Assigner.Typing.DYNAMIC)
+    public static Object[] onEnter(
+        @Advice.AllArguments(typing = Assigner.Typing.DYNAMIC) Object[] arguments,
+        @Advice.FieldValue(value = "retrofit") Retrofit retrofit) {
+      CallDepth callDepth = CallDepth.forClass(InfluxDBImpl.class);
       if (callDepth.getAndIncrement() > 0) {
-        return;
+        return new Object[] {arguments};
```
[the remark this answers is not in the document]
Comment (LikeTheSalad) [reply instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java]: I just changed it to returning null and it seems to be working fine 👍

File: instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java (hunk 2 above)
Comment (LikeTheSalad) [reply instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java]: Thanks, I've made the changes.

File: instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java (hunk 3 above)
Comment (LikeTheSalad) [reply instrumentation/influxdb-2.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/influxdb/v2_4/InfluxDbImplInstrumentation.java]: Yes, I just made the change and it's working fine.

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceTransformer.java (hunk 4 above)
Comment (LikeTheSalad) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceTransformer.java]: Thanks, I've made the changes.

**Review by laurit — APPROVED** [review]
